stable_baselines3/common/recurrent/policies.py

```diff
@@ -615,3 +621,191 @@ def __init__(
+    ):
+        if features_extractor_kwargs is None:
+            features_extractor_kwargs = {}
+        # Automatically deactivate dtype and bounds checks
+        if normalize_images is False and issubclass(features_extractor_class, GRUNatureCNNExtractor):
+            features_extractor_kwargs = features_extractor_kwargs.copy()
+            features_extractor_kwargs.update(dict(normalized_image=True))
+
+        if not issubclass(features_extractor_class, RecurrentFeaturesExtractor):
+            base_features_extractor = features_extractor_class(observation_space, **features_extractor_kwargs)
+
+            features_extractor_class = GRUWrappedFeaturesExtractor
+            new_features_extractor_kwargs = dict(base_extractor=base_features_extractor)
+            if "features_dim" in features_extractor_kwargs:
+                new_features_extractor_kwargs["features_dim"] = features_extractor_kwargs["features_dim"]
+            features_extractor_kwargs = new_features_extractor_kwargs
+            print(features_extractor_class, features_extractor_kwargs)
+
+        super().__init__(
+            observation_space,
+            action_space,
+            lr_schedule,
+            net_arch,
+            activation_fn,
+            ortho_init,
+            use_sde,
+            log_std_init,
+            full_std,
+            use_expln,
+            squash_output,
+            features_extractor_class,
+            features_extractor_kwargs,
+            share_features_extractor,
+            normalize_images,
+            optimizer_class,
+            optimizer_kwargs,
+        )
+
+    def recurrent_initial_state(
+        self, n_envs: Optional[int] = None, *, device: Optional[th.device | str] = None
+    ) -> RecurrentState:
+        return self.features_extractor.recurrent_initial_state(n_envs, device=device)
+
+    def _recurrent_extract_features(
+        self, obs: TorchGymObs, state: RecurrentState, episode_starts: th.Tensor
+    ) -> Tuple[th.Tensor, RecurrentState]:
+        if not self.share_features_extractor:
+            raise NotImplementedError("Non-shared features extractor not supported for recurrent extractors")
+
+        preprocessed_obs = preprocess_obs(obs, self.observation_space, normalize_images=self.normalize_images)  # type: ignore
+        return self.features_extractor(preprocessed_obs, state, episode_starts)
+
+    def forward(  # type: ignore[override]
+        self,
+        obs: TorchGymObs,
+        state: RecurrentState,
+        episode_starts: th.Tensor,
+        deterministic: bool = False,
+    ) -> Tuple[th.Tensor, th.Tensor, th.Tensor, RecurrentState]:
+        """Advances to the next hidden state, and computes all the outputs of a recurrent policy.
+
+        In this docstring the dimension letters are: Time (T), Batch (B) and others (...).
+
+        :param obs: shape (T, B, ...) the policy will be applied in sequence to all the observations.
```

Comment:
Would it be a good idea to add asserts about these shape invariants (ie, obs.shape[0] == episode_starts.shape[0]), or is it more appropriate to rely on math operations failing if those invariants are broken?

Same question applies for `get_distribution` and `predict_values`.

Reply:
I thought it'd give nice error messages, but it's pretty involved to actually check. Then you need to remember to call it under every one of these functions when you instantiate a subclass.

I don't think it's worth it.

Attached is the patch:
[check_recurrent_inputs.patch](https://github.com/AlignmentResearch/learned-planners-stable-baselines3/files/12897269/check_recurrent_inputs.patch)

Reply:
```patch
diff --git a/stable_baselines3/common/recurrent/policies.py b/stable_baselines3/common/recurrent/policies.py
index 19a50a14..b665d994 100644
--- a/stable_baselines3/common/recurrent/policies.py
+++ b/stable_baselines3/common/recurrent/policies.py
@@ -9,7 +9,7 @@ from torch import nn
 from stable_baselines3.common.distributions import Distribution
 from stable_baselines3.common.policies import ActorCriticPolicy
 from stable_baselines3.common.preprocessing import preprocess_obs
-from stable_baselines3.common.pytree_dataclass import tree_flatten
+from stable_baselines3.common.pytree_dataclass import TensorTree, tree_flatten, tree_map
 from stable_baselines3.common.recurrent.torch_layers import (
     ExtractorInput,
     GRUNatureCNNExtractor,
@@ -136,6 +136,21 @@ class BaseRecurrentActorCriticPolicy(ActorCriticPolicy, Generic[RecurrentState])
         """
         ...
 
+    @abc.abstractmethod
+    def _check_recurrent_inputs_congruent(
+        self, observation: TensorTree, state: RecurrentState, episode_starts: th.Tensor
+    ) -> Tuple[int, int]:
+        (time, batch) = episode_starts.shape
+
+        def check_obs_shape(x) -> th.Tensor:
+            if x.shape[:2] != episode_starts.shape:
+                raise ValueError(f"Shape of observation element should be ({time=}, {batch=}, ...), is {x.shape}")
+            return x
+
+        tree_map(check_obs_shape, observation)
+        # Purposefully don't check state here, leave it to specialized classes
+        return (time, batch)
+
     def predict(  # type: ignore[override]
         self,
         obs: TorchGymObs,
@@ -344,6 +359,20 @@ class RecurrentActorCriticPolicy(BaseRecurrentActorCriticPolicy):
             (th.zeros(shape, device=device), th.zeros(shape, device=device)),
         )
 
+    def _check_recurrent_inputs_congruent(
+        self, observation: TensorTree, state: ActorCriticStates[LSTMRecurrentState], episode_starts: th.Tensor
+    ) -> Tuple[int, int]:
+        (time, batch) = super()._check_recurrent_inputs_congruent(observation, state, episode_starts)
+        LSTM_STATE_SHAPE = (self.lstm_actor.num_layers, batch, self.features_dim)
+
+        def _check_state_shape(x: th.Tensor) -> th.Tensor:
+            if x.shape != LSTM_STATE_SHAPE:
+                raise ValueError(f"The shape of the recurrent state is {x.shape=}, but it should be {LSTM_STATE_SHAPE}.")
+            return x
+
+        tree_map(_check_state_shape, state)
+        return (time, batch)
+
     # Methods for getting `latent_vf` or `latent_pi`
     def _recurrent_latent_pi_and_vf(
         self, obs: TorchGymObs, state: ActorCriticStates[LSTMRecurrentState], episode_starts: th.Tensor
@@ -689,6 +718,12 @@ class RecurrentFeaturesExtractorActorCriticPolicy(BaseRecurrentActorCriticPolicy
     ) -> RecurrentState:
         return self.features_extractor.recurrent_initial_state(n_envs, device=device)
 
+    def _check_recurrent_inputs_congruent(
+        self, observation: TensorTree, state: ActorCriticStates[LSTMRecurrentState], episode_starts: th.Tensor
+    ) -> Tuple[int, int]:
+        # TODO implement a similar method for RecurrentFeaturesExtractor
+        return super()._check_recurrent_inputs_congruent(observation, state, episode_starts)
+
     def _recurrent_extract_features(
         self, obs: TorchGymObs, state: RecurrentState, episode_starts: th.Tensor
     ) -> Tuple[th.Tensor, RecurrentState]:

```

Reply:
SGTM either way then. Agreed that the inheritance and abstract methods makes this pretty awkward to do. Deferring to you on what seems like the least bad option.